### PR TITLE
Initial Work

### PR DESCRIPTION
**Pull request by dafuga**

### CODE REVIEW

File: test/index.ts [hunk 1]
```diff
@@ -1,9 +1,36 @@
...
+        sinon.stub(require('bip39'), 'generateMnemonic').returns(words)
+    })
+
+    test('generate', function () {
+        const mnemonicSeed = lib.MnemonicSeed.generate();
+
+        assert.equal(mnemonicSeed.words, words)
+    })
+
+    suite('derivePrivateKey', function () {
```
Comment (jnordberg) [inline test/index.ts:22]: Add a test that verifies that seeds generated by a ledger maps to the public key provided by the ledger eos app

File: src/index.ts [hunk 2]
```diff
@@ -5,9 +5,54 @@
  * @param hard Set to true for super hard problem.
  * @returns The solution.
  */
-export function maths(n: number, p: number, hard = false) {
-    while (hard) {
-        n = n * p
+import {generateMnemonic, mnemonicToSeed} from 'bip39'
+import BIP32Factory from 'bip32'
+import * as ecc from 'tiny-secp256k1'
```
Comment (jnordberg) [inline src/index.ts:10]: We should avoid pulling in an additional ecc lib if possible, core uses elliptic.js

Comment (dafuga) [reply src/index.ts:10]: I am trying to use `elliptic` instead, but I keep on getting a `ecc.isPoint is not a function` error. Do you know if `elliptic` has a method that checks if a point is on the elliptic curve?

File: test/index.ts (hunk 1 above)
Comment (dafuga) [reply test/index.ts:22]: Hmm, not sure I am following. Are you saying that you want me to generate a public key using the seed (something like [this](https://github.com/iancoleman/bip39/blob/23f393acfba3f8697f42249965f4acfebae918a2/src/js/eos-util.js#L1-L7)) and compare it to the public key that is returned by core?

Comment (dafuga) [reply test/index.ts:22]: Just added a test for this 👍

**Comment by dafuga** [comment]
Thanks for the feedback @jnordberg ! I added the test that you suggested. As discussed during our dev call, I think we can keep the ECC and BIP39 libraries around for now and potentially get rid of them in a future PR.

File: test/index.ts [hunk 3]
```diff
@@ -1,9 +1,44 @@
...
-suite('index', function () {
-    test('maths', function () {
-        assert.equal(lib.maths(1, 2), 3)
+suite('MnemonicSeed', function () {
+    const words = 'enrich cycle essay coast asthma shadow heavy business inject volcano era bamboo grid avoid magic fortune first toilet various grit river cash luggage fan';
+    const ledgerWords = 'choice defense mushroom pioneer crime kite flock rebel point submit treat earn play excess fence depth whip unveil easily pluck will talk brand grain';
+
+    suiteSetup(function () {
+        // eslint-disable-next-line @typescript-eslint/no-var-requires
+        sinon.stub(require('bip39'), 'generateMnemonic').returns(words)
```
Comment (jnordberg) [inline test/index.ts:13]: Why is this stubbed out? Seems like a critical thing we want to test

File: src/index.ts [hunk 4]
```diff
@@ -5,9 +5,74 @@
...
+        this.mnemonicWords = mnemonicWords
     }
-    return n ^ p
+
+    public get words(): string {
+        return this.mnemonicWords
+    }
+
+    public get hex(): Promise<string> {
+        return new Promise((resolve, reject) => {
```
Comment (jnordberg) [inline src/index.ts]: Manually creating a new promise should only be done when bridging callback-style APIs to promises. This should be an async function

File: src/index.ts (hunk 2 above)
Comment (jnordberg) [reply src/index.ts:10]: `ec('secp256k1').verify(..)` should do it I think

File: src/index.ts [hunk 5]
```diff
@@ -5,9 +5,74 @@
...
+
+        if (!bip32Interface?.privateKey) {
+            throw new Error('Failed to derive private key')
+        }
+
+        const privateKeyBuffer = Bytes.from([0x80]).appending(
+            Bytes.from(bip32Interface.privateKey!)
+        )
+
+        const checksum = dsha256Checksum(privateKeyBuffer.array)
```
Comment (jnordberg) [inline src/index.ts]: This isn't needed, just call the private key constructor with the data instead.

`new PrivateKey(KeyType.from('K1'), Bytes.from(privateKeyData)` (no 0x80 padding)

**Review by jnordberg — CHANGES_REQUESTED** [review]
I'm not able to run the browser tests:

```
$ make browser-test

tests → test/browser.html...
[!] (plugin commonjs--resolver) Error: Unexpected character '' (Note that you need plugins to import files that are not JavaScript)
node_modules/tiny-secp256k1/lib/secp256k1.wasm (1:0)
generateInt32./validate_error.js``````~~~~``~<	./rand.js
   ^
2: throwError`_
3:
Error: Unexpected character '' (Note that you need plugins to import files that are not JavaScript)
    at error (/Users/jnordberg/Development/mnemonic-seeds/node_modules/rollup/dist/shared/rollup.js:198:30)
    at Module.error (/Users/jnordberg/Development/mnemonic-seeds/node_modules/rollup/dist/shared/rollup.js:12555:16)
    at Module.tryParse (/Users/jnordberg/Development/mnemonic-seeds/node_modules/rollup/dist/shared/rollup.js:12932:25)
    at Module.setSource (/Users/jnordberg/Development/mnemonic-seeds/node_modules/rollup/dist/shared/rollup.js:12837:24)
    at ModuleLoader.addModuleSource (/Users/jnordberg/Development/mnemonic-seeds/node_modules/rollup/dist/shared/rollup.js:22293:20)
```

File: test/index.ts (hunk 3 above)
Comment (dafuga) [reply test/index.ts:13]: It's already tested in the library so I figured that we could stub it.

File: src/index.ts (hunk 2 above)
Comment (dafuga) [reply src/index.ts:10]: I don't think those methods are compatible. After mapping `isPoint` to `ec('secp256k1').verify(..)` I get a `TypeError: this._truncateToN is not a function` error.

File: test/index.ts (hunk 3 above)
Comment (dafuga) [reply test/index.ts:13]: I guess I could remove the stub though, the tests should still work.

Comment (dafuga) [reply test/index.ts:13]: Cause I need a way to always obtain the same words so I can make sure that our methods are working 😄

**Comment by dafuga** [comment]
@jnordberg thanks for the review! I added `@rollup/plugin-wasm` and I'm no longer getting that error. I also addressed some of your other feedback. Let me know if you would change anything else!

**Comment by dafuga** [comment]
@jnordberg Any thought on this? 😄

**Comment by dafuga** [comment]
I'm gonna go ahead and merge this since it's functional. If we have ideas on how to make it better we can open a new PR.

**Pull request merged**